### ISSUE TITLE
Effort tracking

### ISSUE THREAD

**Issue by jahto** [Flyspray/flyspray#96]
It seems that both I and peterdd have been working on effort tracking on slightly different viewpoints. Let's have a discussion what we want to actually implement.

What I really need to have is to change estimated effort from hours to seconds, like actual effort. An hour is too coarse-grained for really small tasks. I and Peterdd seem to agree that 24 hours != 1 day. I used only hours:minutes for the display form, peterdd implemented manday (8 hours == manday). I'm fine with that too, although in theory I think that manhours have in reality almost nothing to do with any kind of days. The editable form, in my opinion, should always be H+(:mm)? to be easily parseable.

My suggestion:
- we come up with a resolution of what we implement
- peterdd makes a pull request, he's got a lot of other important things too to merge
- after that, I do a merge, fix the resulting mess and implement what we agreed upon, I've got time and there's nothing else that I see really important for 1.0

**Comment by Psychokiller1888** [comment]
Hours to second seems monstuous... To minutes would be maybe something better... What about an option to chose from? I mean, make them available to what the team needs.... I won't be able to do anything with seconds, I need days in efforts...

So my opinion is to add an option in the project pannel to chose the unit of effort tracking

**Comment by jahto** [comment]
I'd say it's just easier to store both estimated and actual effort as seconds in database, then both can share some functionality for parsing and formatting, but the display and editable formats could of course be choosable by the user. And as an afterthought to peterdd's manday, it could be a setting per project, someones manday is 8 hours, someone elses 7.5, etc. And 0 could mean 'just show plain hours'. Does this sound reasonable?

**Comment by Psychokiller1888** [comment]
Storing is anyway better in seconds, and i wouldn't do anything else. Makes it easy to display whatever you want then. The as said, a project option for the display format and we're good to go

**Comment by jahto** [comment]
I'll start to plan how to implement the functionality so that it fits both my, your and everybody elses needs. But I'm not going to push anything to github until peterdd and possibly Jordan has had his say on this issue, and not until peterdd has made a pull request that hopefully will get accepted. There WILL be conflicts that must be manually resolved, we've been messing with the same area of code.

**Comment by peterdd** [comment]
Making man-days length configurable per project sounds a bit overengineered to me. It is not a requirement for me, I just added man-day display for getting an overview of the estimated project effort. I'm still unsure in which situations (task list, task details view, milestones, ..) which display (hh:mm, only hours, man-days or a mixed display) is best. 

Well, in fact I just looked at https://de.wikipedia.org/wiki/Personenstunde
where is stated that in europe a man-day is calculated like 8 work hours.

How about to let the user choose if he want also a display for man-days or just show the hh:mm ?

Making estimated effort also editable in hh:mm form as jahto needs sound valid, for me are hour steps enough.

But internal (php/database) all effort handling/calculation should be in seconds to avoid confusion. (done effort is in secondes too)

My current fork isn't stable enought to merge with all that stuff that is to do to implement the anti csrf system and session data storage changes. Are you the same opinion that lurking project_id, passhash, task_type in usercookies isn't the best solution? I think it lowers usability because you cannot handle parallel 2 projects in 2 tabs of the browser. Whenever you change something in one tab it interferes the other tab.

Is the oauth stuff in working condition? Will something get broken if we remove user_id and passhash token cookies?

**Comment by jahto** [comment]
In Finland, we use in reality anything between 7 and 7.5 hours... So it needs at least to be configurable. Per project might be overengineering, but I've got it already half way implemented, so I think it will stay per project setting.

If peterdds' fork is not stable enough yet, l'll keep on pushing to my fork, branch effort_tracking_fixes so you can at least see which direction I'm going to and make comments if necessary. Keep in mind it's still work-in-progress. But I will make a pull request only after peterdds' fork gets merged, if he can get it finished before feature freeze, and after resolving manually the resulting conflicts.

About the anti crsf system and session data storage changes I have absolutely no opinions with the exception that it's best left to someone who knows what he's doing and I'm not that person.

**Comment by Psychokiller1888** [comment]
Haaaa, so you are finish :) My wife was correct. She saw your name and directly asked: "Is he finish?" Because she is finish...

In Switzerland we have 8 hours manday.

What i'd say is better make configurable and good directly than rework later. Storing in second is a must, then display how you want. Some may need hours, others minutes... So it should be configurable. And I think that is a project setting as you could host different projects with different requirements...

**Comment by jahto** [comment]
Yes, I am finnish, note that it's spelled with two n's. By coincidence, I too have some relatives in Switzerland, mostly living somewhere near Aargau. If you are brave enough you can try some day to say to her when coming home "terppa taas puluseni", if she has a good sense of humour. If not, she might hit you...

Funny that currently the most active people working on Flyspray all europeans. A finn, a swiss and a german.

But back to the topic. You can see what I'm planning to do on github. Tomorrow I'll probably add switch/case constructs for different formats to effort::SecondsToString and effort::SecondsToEditString and check that the parsing (at least in some cases) works. After that, you really should comment if you don't find it fitting your needs.

**Comment by jahto** [comment]
I'm almost done with this one, but begun to think about a few enhancements more. Like adding a few display formats more, or rather splitting existing ones to several versions ("plain hours", "hours with one decimal", "plain working days", "working days with one decimal", "plain working days plain hours", "plain working days hours with one decimal") so everyone really gets the resolution he needs (someone will eventually want two or more decimals but that someone must implement those ones him/serself, I say this would be already more than enough) and splitting estimated effort and actual effort formats, the first one could be used to give just a rough estimate, the last one could be used for example for accounting and in that case needs to be quite precise. Would take only a few hours more.

**Comment by Psychokiller1888** [comment]
Great! That's what we were talking about, making it fit to everyone needs!

**Comment by peterdd** [comment]
Well, is there a really need for that?

If you really want it that flexible its maybe better to provide a format string input field for output displaying instead of a selection.


That is what I know from Joomla for example, where Date and Datetimes can be formatted by php format strings in event calender extensions. Or Worpress for the URLpath construction..

And keep an eye on the correct roundings on effort output. e.g. 
a database value of 6600:
IMHO valid displays for that are:
01:50
110 min
1 h 50 min
2 h (rounding up/ceil() if you are only interested in full hours for billing)
1,9 h (hours with 1 decimal round up/ceil() for billing) , also if ',' or '.' is the decimal divider is localization dependend, in Germany we use ',' for that and '.' for 1000's dividers..
1 man-day/Manntag if you bill only whole days.
0,3 man-days/Manntage if you want 1 decimal and calc with 8 hours.

Postpone it for Flyspray 1.1 ?

More important (and not so easy to done "right") is to make "estimated effort" and "done effort" display on the task list view and task view if tasks have sub tasks which have sub tasks which have sub tasks ... (recursive)

For a 1.0 release I think it is ok if the users sees in the task list view on the 'estimated effort' and 'done effort' that this task has sub tasks which also can have effort values which must be considered. We can use the info we display on the parent column for that too.

On the task detail view there should be a note/sign the subtasks exists which have estim. and done effort too.

For the native english speakers: is 'actual_effort' the right term for effort, that is done (work time, even if there is no valueable result)
(It has nothing to do with the 'progress' in percent of the task)

**Comment by jahto** [comment]
I like predefined formats because I suspect everything that the user can define and input himself... And yes, I'll really have to think when to use floor/round/ceil. I don't think that his has to postponed to 1.1, as I said, it's only a few hours of effort more. What is too much effort for 1.1 is the recursive part, unfortunately.

**Comment by peterdd** [comment]
Just as reminder for effort calculations on nested tasks in Flyspray 1.1:

There are technics to get infos from tree like data with one sql-query on sql-databases which don't support recursive queries out of the box. Well - if the db structure is good for that.

The current  category_id, lft, rgt structure of category table seems good for that.
while a simple tast_id, parent_id doesn't fit well queries over a tree.

There is literature on the net for that topic.

**Comment by jahto** [comment]
In my opinion, effort tracking has now enough features and is stable enough to be considered for merging into master. I'll make a pull request late today if it's ok for everyone or on friday if you find still something (after fixing it). The rest of possible additional features should wait for 1.1 or we will never get 1.0 released.

**Comment by peterdd** [comment]
Ok, if you think it is stable it is good if you merge.

I have the same opinion. Feature freeze and then only allow fixing bugs for 1.0. 
(php5.2, php5.3, php5.4, php5.5, php5.6, mysql/mariadb, postgres,...)
Review and solve/close bug issues on bugs.flyspray.org

You both worked on the installer and upgrader. Does the upgrader also handles the situation, when somebody installed a github dev version? (for getting db changes from January 2015 to now from dev to 1.0)

The upgrade info on the README.md must be adapted for composer usage in 1.0.

I have not much time to do anything on flyspray until middle next week.

**Comment by Psychokiller1888** [comment]
Yeah, feature freeze mid march. Readmes will need review and rework, but we still have time :)

**Comment by jordanmendler** [comment]
There is somewhere I already had an updated step by step on how to install
with composer. Can't remember if on the website or readme. Let me know if
you don't see it and I will track it down

_________________

Jordan Mendler
President & Chief Technology Officer
The Veloz Group | www.thevelozgroup.com

**Comment by jahto** [comment]
Re "Does the upgrader also handles the situation, when somebody installed a github dev version?": upgrader seems to handle it now quite well. I do a drop database and restore from a db dump almost daily, from version 0.9.9.7, and also do an upgrade between development versions when I switch my working branch from anything where database schema is the current dev version to my effort_tracking_fixes branch. I guess I could do some more testing on weekend and fix any issues I might still find.

**Comment by peterdd** [comment]
Great!
I added/updated some newer added db fields according to dev version by hand on a flyspray in 'semi'-production environment and think many others are having their flyspray just using a github dev version because the last official release so so outdated. 
Should not be to complicated as there was only adding fields or adding index/unique keys to the db, but no renaming or deletion of fields.

**Comment by Psychokiller1888** [comment]
I think the OP is closed, effort tracking is done
